### ISSUE TITLE
Huge performance degradation between 3.2 and 3.6.3

### ISSUE THREAD

**Issue by RobSwDev** [microsoft/testfx#4132]
## Steps To Reproduce
I can't share our code base, but I did do some poking around with windbg:

It looks like all the test class instances are held onto permanently. 
Each of our test classes has member variables that hold several MB of data (This may be dubious practice, but it's how our tests been built)
If these instances are not garbage collected then memory keeps growing.

`ExecutionContextService.InstancesExecutionContexts`  in Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices seems to be the culprit. For whatever reason, none of the test instances are being cleaned up - I still have 1000s of test instances rooted.

## Expected behavior
Performance should not degrade

## Actual behavior
A set of tests that previously took 5 minutes now takes 1 hour.

## Additional context
Command used to run tests has a rather large test case filter, in case that is relevant
```
vstest.console test.dll 
/Testcasefilter:"(TestCategory!=A & TestCategory!=B & .. & TestCategory!=Z)" 
/Settings:.runsettings
/inIsolation 
/Logger:trx;LogFileName="Log.trx" 
/ResultsDirectory:"c:\temp"
```

**Comment by RobSwDev** [comment]
Our RunSettings file contains the following
``` xml
  <RunConfiguration>
    <MaxCpuCount>1</MaxCpuCount>
  </RunConfiguration>
```
I imagine the reasons for this setting are lost in the mists of time.
In a sample to repro the issue, I've found this makes a huge difference.

**Comment by nohwnd** [comment]
Hello, thank you for spelling out the details of what caused this, this will make fixing it way easier I hope.

I tried to make a project to reproduce the regression, but I don't see it being slower. Can you tell me more about your project, or provide one that shows the behavior?

Are you compiling as .NET or .NET Framework ?
What are the parallelization settings you use for mstest (the maxcpu setting you shared above is for vstest, mstest uses assembly: Parallelization attribute, or a specific runsetting in <MSTest> section)

**Comment by RobSwDev** [comment]
[MsTestAdapterRegression.zip](https://github.com/user-attachments/files/17981053/MsTestAdapterRegression.zip)

I've managed to create a sort of repro. Our real tests are big and gnarly, and there may be more going on that I've missed.
But the attached project illustrates a slowdown: 
* 2.5 mins using 3.2.0
* 11.4 mins using 3.6.3

To repro: build in release, and then run (from each project's respective bin\release folder)
``` powershell
vstest.console .\Adapter3.2.dll "/settings:Test.runsettings"
vstest.console .\Adapter3.6.dll "/settings:Test.runsettings"
```

[We don't help ourselves with the way our tests have evolved - I've discovered that we're forcing GC twice per test, unnecessarily setting MaxCpuCount to 1, etc.]

**Comment by RobSwDev** [comment]
We're using Net8,0.

Our CI pipeline currently runs vstest.console with a runsettings file.
Would we be better off running `dotnet vstest`?

Does the framework now cache every instance created of a test class? Or does it depend on how we use it?

**Comment by nohwnd** [comment]
Thanks for the repro I will have a look as soon as I can. 

You would not be better off using dotnet vstest. All dotnet test, dotnet vstest and vstest.console eventually call to vstest.console, there are some minor differences but the regression is directly in the test framework. We started holding on to the test contexts to be able to restore them for specific situations, like if the user sets locale for the test.

To improve performance of your tests you might also consider moving to the new testing.platform https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-intro?tabs=dotnetcli where our customers see 30% increase in speed. But as you say if lot of those problems are self-inflicted you might not see any bump in performance.

**Comment by RobSwDev** [comment]
I was able to look into this a bit further, and found the "cause", and a bizarre workaround.

It seems, if I specify a `TestCleanupTimeout` value in the RunSettings, then there is no leak.
If I don't, then the code that is supposed to remove the contexts is not called.

In `FixtureMethodRunner.RunWithTimeoutAndCancellation`, if the parameter `timeoutInfo` is null, then context cleanup is skipped.
But if it is non-null, then execution proceeds to `RunWithTimeoutAndCancellationWithThreadPool`, which calls `ExecutionContextService.RunActionOnContext`, which calls `CleanupExecutionContext` and all is good.

Setting TestCleanupTimeout seems to be a workaround.


The attached project illustrates this: both via memory usage, and via a reflection hack to get the internal dictionary.
If you run the tests with settings file `Test.runsettings`, it takes over a minute, and ~ 20GB
If you run them with settings file `WithTestCleanupTimeout.runsettings`, it takes 10 seconds and less than 2GB
(performance obviously depends on e.g. available ram, etc)



[AdapterLeak.zip](https://github.com/user-attachments/files/18693174/AdapterLeak.zip)

**Comment by nohwnd** [comment]
That's crazy. Will try to see if I can repro, zince I was not able to see that behavior before.

**Comment by nohwnd** [comment]
oh yeah, takes 30s and 20gb with the test.runsettings, and 10s and 2gb with WithTestCleanupTimeout.runsettings (or without any runsettings -  which is why I did not see the problem before).

**Comment by nohwnd** [comment]
Interestingly 48s in the mtp playground.

**Comment by nohwnd** [comment]
25s with the server gc setting.

**Comment by nohwnd** [comment]
Yeah so the problem comes from here where, action without timeout will be invoked as pure action.

https://github.com/microsoft/testfx/blob/c1d2aaf10d9d4b29adaccd4b2080c3dbe4d8088c/src/Adapter/MSTest.TestAdapter/Helpers/FixtureMethodRunner.cs#L36

This action does not call into ExecutionContextService, and so it does not perform the cleanup.

 But an action with timeout will be invoked via RunWithTimeoutAndCancellationWithThreadPool (e.g)

https://github.com/microsoft/testfx/blob/c1d2aaf10d9d4b29adaccd4b2080c3dbe4d8088c/src/Adapter/MSTest.TestAdapter/Helpers/FixtureMethodRunner.cs#L69

this does call into  ExecutionContextService.RunActionOnContext and performs the cleanup.

This can be reproed in our Playground, the time differences are not so big, but setting breakpoints into https://github.com/microsoft/testfx/blob/c1d2aaf10d9d4b29adaccd4b2080c3dbe4d8088c/src/Adapter/MSTestAdapter.PlatformServices/Services/ExecutionContextService.cs#L71

you will see the cleanup is never invoked when the test.runsettings (provided above are used).

(Maybe there is also a backwards problem where the context is not captured.)

**Comment by nohwnd** [comment]
repro branch with code in our playground for easy debugging: https://github.com/microsoft/testfx/compare/repro4132?expand=1

**Comment by Youssef1313** [comment]
@nohwnd It's more complicated that this, I think.
If there is no TestCleanup at all, we will never clear the caches. In that case, I don't think the workaround for `TestCleanupTimeout` is going to work.

**Comment by nohwnd** [comment]
Too bad. But in that case the logic must change somehow. To indicate that there is a need to clean up our internal state, and not be tied to the user providing cleanup.

**Comment by Youssef1313** [comment]
@nohwnd I agree.
